functions/authenticate.js

```diff
@@ -0,0 +1,123 @@
+			return;
+		}
+		
+		var responseBody = {
+			StackId: event.StackId,
+			RequestId: event.RequestId,
+			LogicalResourceId: event.LogicalResourceId,
+			PhysicalResourceId: context.logStreamName,
+			Status: status,
+			Reason: getReason(err) + " See details in CloudWatch Log: " + context.logStreamName,
+		};
+
+		console.log("RESPONSE:\n", responseBody);
+		var json = JSON.stringify(responseBody);
+		var parsedUrl = url.parse(event.ResponseURL);
+		var options = {
+			hostname: parsedUrl.hostname,
+			port: 443,
+			path: parsedUrl.path,
+			method: "PUT",
+			headers: {
+				"content-type": "",
+				"content-length": json.length
+			}
+		};
+
+		var request = https.request(options, function(response) {
+			console.log("STATUS: " + response.statusCode);
+			console.log("HEADERS: " + JSON.stringify(response.headers));
+			context.done(null, null);
+		});
+
+		request.on("error", function(error) {
+			console.log("sendResponse Error:\n", error);
+			context.done(error);
+		});
+
+		request.on("end", function() {
+			console.log("end");
+			resolve();
+		});
+		
+		request.write(json);
+		request.end();
+	});
+}
+
+async function getNewAccessToken(authUrl, client_id, user_key) {
+	let apiAccessAuthOptions = url.parse(authUrl);
+    let apiAccessAuthBody = JSON.stringify({ 
+        grant_type: 'refresh_token',
+        client_id, 
+        refresh_token: user_key,
+    });
+
+    apiAccessAuthOptions.method = 'POST';
+    apiAccessAuthOptions.headers = {
+        'Content-Type': 'application/json',
+        'Content-Length': apiAccessAuthBody.length
+    };
+
+	return new Promise((resolve, reject) => {
+        let apiAccessAuthReq = https.request(apiAccessAuthOptions, function (res) {
+            res.setEncoding('utf8');
+            res.on('data', function (res) {
+                let tokensObj = JSON.parse(res);
+                if(!tokensObj.access_token) {
+					throw 'access_token not found';
+                }
+                
+                resolve(tokensObj.access_token);
+            });
+        });
+    
+        apiAccessAuthReq.on('error', reject);
+        apiAccessAuthReq.write(apiAccessAuthBody)
+        apiAccessAuthReq.end();
+    });
+}
+
+exports.handler = async (event, context) => {
+	let client_id;
+	let user_key;
+	
+	try {
+		client_id = await getSecret(process.env.api_access_key_client_id_secret_id);
+		user_key = await getSecret(process.env.api_access_key_user_key_secret_id);
+	} catch (err) {
+		await sendResponse(event, context, 'FAILED', err);
+	}
+    
+	try {
+		const access_token = await getNewAccessToken(process.env.api_access_auth_url, client_id, user_key);
+		await updateSecret(process.env.access_token_secret_id, access_token);
+		await sendResponse(event, context, 'SUCCESS');
+	} catch (err) {
+		await sendResponse(event, context, 'FAILED', err);
+		appInsightsClient.trackException({exception: err, measurements: {
+            accountName,
```

Comment:
Can we also include StackId and any other relevant information?
If we can get a ContactFlowInstanceId, that can also act as a correlation Id for all events and exceptions for that run of the contact flow.

Reply:
Have include the lambda name instead. This was needed by Justin's requirements for dashboards.